### PR TITLE
Add Intent Detection Node

### PR DESCRIPTION
**Pull request by derkmed**
This PR should be stacked on https://github.com/PTG-Kitware/angel_system/pull/295. Run `tmuxinator start eval_vocalized_intent_detection` to demo.

### CODE REVIEW

File: tmux/eval_vocalized_intent_detection.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            -p debug_mode:=True
+  - intent_detection:
+      layout: even-vertical
+      panes:
+        - base_intent_detection: ros2 run angel_system_nodes base_intent_detector --ros-args
+           -r __ns:=${ROS_NAMESPACE}
+           -p utterances_topic:=utterances_topic
+           -p expect_user_intent_topic:=expect_user_intent_topic
+           -p interp_user_intent_topic:=interp_user_intent_topic
+        - gpt_intent_detection: ros2 run angel_system_nodes gpt_intent_detector --ros-args
```
Comment (joshanderson-kw) [inline tmux/eval_vocalized_intent_detection.yml:92]: Do both of these nodes actually need to run? Or would just one of them need to run in order to publish intents?

File: ros/angel_system_nodes/angel_system_nodes/gpt_intent_detector.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+# The following are few shot examples when prompting GPT.
+FEW_SHOT_EXAMPLES = [
+    {"utterance": "Go back to the previous step!", "label": "prev_step."},
+    {"utterance": "Next step, please.", "label": "next_step"},
+    {"utterance": "How should I wrap this tourniquet?", "label": "inquiry"},
+    {"utterance": "The sky is blue", "label": "other"},
+]
+
+
+class GptIntentDetector(BaseIntentDetector):
```
Comment (joshanderson-kw) [inline ros/angel_system_nodes/angel_system_nodes/gpt_intent_detector.py:23]: Are we planning to use this for the November demo?

File: ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+        classification, confidence = _tiebreak_intents(intents, confidences)
+        classification = colored(classification, "light_green")
+        return classification, confidence
+
+    def publish_msg(self, utterance, intent, score):
+        """
+        Handles message publishing for an utterance with a detected intent.
+        """
+        intent_msg = InterpretedAudioUserIntent()
```
Comment (joshanderson-kw) [inline ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py:149]: Also fill in ROS header for this message.

File: ros/angel_system_nodes/angel_system_nodes/gpt_intent_detector.py (hunk 2 above)
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/gpt_intent_detector.py:23]: We can  use either the BaseIntentDetector (keyphrase matching) or  GptIntentDetector. I decoupled them such  that we can easily replace one with the other based on our use cases.

File: tmux/eval_vocalized_intent_detection.yml (hunk 1 above)
Comment (derkmed) [reply tmux/eval_vocalized_intent_detection.yml:92]: Only one of them  needs to run for intent detection. This .yml runs  both just  for testing purposes.

File: ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        Keyphrase search for intent detection. This implementation does simple
+        string matching to assign a detected label.
+        """
+
+        def _tiebreak_intents(intents, confidences):
+            classification = intents[0]
+            score = confidences[0]
+            if len(intents) > 1:
+                for i, intent in enumerate(intents):
+                    if intent == INTENT_LABELS[2]:
```
[the remark this answers is not in the document]
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py:117]: Yes!

File: ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        )
+        self._interp_publisher = self.create_publisher(
+            InterpretedAudioUserIntent, self._interp_uintent_topic, 1
+        )
+
+        self.message_queue = queue.Queue()
+        self.handler_thread = threading.Thread(target=self.process_message_queue)
+        self.handler_thread.start()
+
+    def listener_callback(self, msg):
```
[the remark this answers is not in the document]
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py:86]: Done.

File: ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        super().__init__(self.__class__.__name__)
+        self.log = self.get_logger()
+
+        # Handle parameterization.
+        parameter_names = [
+            UTTERANCES_TOPIC,
+            PARAM_EXPECT_USER_INTENT_TOPIC,
+            PARAM_INTERP_USER_INTENT_TOPIC,
+        ]
+        set_parameters = self.declare_parameters(
```
[the remark this answers is not in the document]
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py:36]: Done.

File: ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py (hunk 3 above)
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/base_intent_detector.py:149]: Not sure if I filled it out properly, but I think this  should be done.

**Comment by derkmed** [comment]
Linking relevant issue: https://github.com/PTG-Kitware/angel_system/issues/307

**Pull request merged**